lib/spack/spack/spec_list.py

```diff
@@ -144,7 +140,16 @@ def sigilify(arg):
                                 return arg
                             else:
                                 return sigil + arg
-                        return list(map(sigilify, ret))
+
+                        # Apply the sigil if any to current reference
+                        current = self._reference[name].specs_as_yaml_list
+                        current = list(map(sigilify, current))
+
+                        # Recurse on the rest of the entry
+                        ret = [self._expand_references(i) for i in yaml[:idx]]
+                        ret += current
+                        ret += self._expand_references(yaml[idx + 1:])
+                        return ret
                     else:
```

Comment:
Can you:

* wrap everything at this level of indentation in a new function
* do this error checking and fail up front (thereby reducing the indentation level for this new function)
* avoid re-calling `_expand_references` on prior elements of the data structure that have already been processed by this function

? I think these changes would all go towards making this code more readable.

Reply:
I did a complete refactor. I think I addressed your concerns, although I did it completely differently.